apache/trafficcontrol
# Added Traffic Portal page for CDNi config requests

This PR adds an endpoint to TO to GET all of the pending CDNi configuration update requests and a page in TP to view them and approve/deny

<hr/>

## Which Traffic Control components are affected by this PR?
- Documentation
- Traffic Ops
- Traffic Portal

## What is the best way to verify this PR?
- Verify that all TO and TP tests pass
- Verify that all docs build and look good
- 1. Run TO and TP. (Be sure your [TO configuration file ](https://traffic-control-cdn.readthedocs.io/en/latest/admin/traffic_ops.html#cdn-conf) sets the `cdni` property properly)
    2. Navigate to the new `cdni-config-requests` page using the side menu Configuration > CDNi Requests
    3. Manually put CDNi data in the database (test INSERTs for data below)
    4. Submit a configuration update request to `/api/4.0/OC/CI/configuration/`
    5. Verify that the data loads successfully
    6. Click on one of the rows and verify that it takes you to the corresponding request page by ID
    7. Click Approve and verify that the new data is implemented at the `/OC/FCI/advertisement` endpoint
    8. Submit another request, verify it shows up
    9. Deny the new request and verify that the data is NOT updated at the `/OC/FCI/advertisement` endpoint

```sql

INSERT INTO cdni_capabilities ("type", ucdn) VALUES ('FCI.CapacityLimits', 'ucdn1');
INSERT INTO cdni_capabilities ("type", ucdn) VALUES ('FCI.Telemetry', 'ucdn1');
INSERT INTO cdni_footprints (footprint_type, footprint_value, ucdn, capability_id) VALUES ('countrycode', ARRAY['us'], 'ucdn1', 1);
INSERT INTO cdni_footprints (footprint_type, footprint_value, ucdn, capability_id) VALUES ('ipv4cidr', ARRAY['127.0.0.1','127.0.0.2'], 'ucdn1', 1);
INSERT INTO cdni_telemetry (id, "type", capability_id) VALUES ('capacity_metrics', 'generic', 2);
INSERT INTO cdni_telemetry (id, "type", capability_id) VALUES ('request_metrics', 'generic', 2);
INSERT INTO cdni_telemetry_metrics ("name", time_granularity, data_percentile, latency, telemetry_id) VALUES ('capacity', 0, 50, 0, 'capacity_metrics');
INSERT INTO cdni_telemetry_metrics ("name", time_granularity, data_percentile, latency, telemetry_id) VALUES ('requests', 0, 50, 0, 'request_metrics');
INSERT INTO cdni_host_limits (limit_type, maximum_hard, maximum_soft, telemetry_id, capability_id, telemetry_metric, host) VALUES ('requests', 200, 10, 'request_metrics', 1, 'requests', 'example.com');
INSERT INTO cdni_host_limits (limit_type, maximum_hard, maximum_soft, telemetry_id, capability_id, telemetry_metric, host) VALUES ('requests', 20, 15, 'request_metrics', 1, 'requests', 'example2.com');
INSERT INTO cdni_total_limits (limit_type, maximum_hard, maximum_soft, telemetry_id, capability_id, telemetry_metric) VALUES ('egress', 3000, 500, 'capacity_metrics', 1, 'capacity');
```

## PR submission checklist
